providers/vercel/vercelProvider.go

```diff
@@ -0,0 +1,394 @@
+package vercel
+
+/*
+Vercel DNS provider (vercel.com)
+
+Info required in `creds.json`:
+	- team_id
+	- api_token
+*/
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/StackExchange/dnscontrol/v4/models"
+	"github.com/StackExchange/dnscontrol/v4/pkg/diff2"
+	"github.com/StackExchange/dnscontrol/v4/providers"
+	"github.com/miekg/dns"
+	vercelClient "github.com/vercel/terraform-provider-vercel/client"
+)
+
+var features = providers.DocumentationNotes{
+	// The default for unlisted capabilities is 'Cannot'.
+	// See providers/capabilities.go for the entire list of capabilities.
+	providers.CanAutoDNSSEC:          providers.Cannot(),
+	providers.CanGetZones:            providers.Cannot(),
+	providers.CanConcur:              providers.Unimplemented(),
+	providers.CanUseAlias:            providers.Can(),
+	providers.CanUseCAA:              providers.Can(),
+	providers.CanUseDHCID:            providers.Cannot(),
+	providers.CanUseDS:               providers.Cannot(),
+	providers.CanUseDSForChildren:    providers.Cannot(),
+	providers.CanUseLOC:              providers.Cannot(),
+	providers.CanUseNAPTR:            providers.Cannot(),
+	providers.CanUsePTR:              providers.Cannot(),
+	providers.CanUseSOA:              providers.Cannot(),
+	providers.CanUseSRV:              providers.Can(),
+	providers.CanUseSSHFP:            providers.Cannot(),
+	providers.CanUseTLSA:             providers.Cannot(),
+	providers.CanUseDNSKEY:           providers.Cannot(),
+	providers.DocCreateDomains:       providers.Unimplemented(),
+	providers.DocDualHost:            providers.Cannot("Vercel does not allow sufficient control over the apex NS records"),
+	providers.DocOfficiallySupported: providers.Cannot(),
+}
+
+// vercelProvider stores login credentials and represents and API connection
+type vercelProvider struct {
+	client   vercelClient.Client
+	apiToken string
+	teamID   string
+
+	createLimiter *rateLimiter
+	updateLimiter *rateLimiter
+	deleteLimiter *rateLimiter
+	listLimiter   *rateLimiter
+}
+
+// uint16Zero converts value to uint16 or returns 0, use wisely
+//
+// Vercel's Go SDK implies int64 for almost everything, but since Vercel doesn't actually
+// implement their own NS and instead uses NS1 / Constellix (previously), we'd assume if
+// TTL and Priority are int64, they are in fact uint16 and otherwise be rejected by upstream
+// providers. Under this assumption, we'd convert int64 to uint16 as wells.
+func uint16Zero(value interface{}) uint16 {
```

Comment:
```suggestion
func uint16Zero[T ~float64 | ~uint16](value T) uint16 {
```

Generic could be used here.